hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestChunkInputStream.java

```diff
@@ -0,0 +1,198 @@
+    }
+  }
+
+  /**
+   * Match readData with the chunkData byte-wise.
+   * @param readData Data read through ChunkInputStream
+   * @param inputDataStartIndex first index (inclusive) in chunkData to compare
+   *                            with read data
+   * @param inputDataEndIndex last index (exclusive) in chunkData to compare
+   *                          with read data
+   */
+  private void matchWithInputData(byte[] readData, int inputDataStartIndex,
+      int inputDataEndIndex) {
+    for (int i = inputDataStartIndex; i < inputDataEndIndex; i++) {
+      Assert.assertEquals(chunkData[i], readData[i - inputDataStartIndex]);
+    }
+  }
+
+  /**
+   * Seek to a position and verify through getPos().
+   */
+  private void seekAndVerify(int pos) throws Exception {
+    chunkStream.seek(pos);
+    Assert.assertEquals("Current position of buffer does not match with the " +
+        "seeked position", pos, chunkStream.getPos());
+  }
+
+  @Test
+  public void testFullChunkRead() throws Exception {
+    byte[] b = new byte[CHUNK_SIZE];
+    chunkStream.read(b, 0 , CHUNK_SIZE);
+
+    matchWithInputData(b, 0, CHUNK_SIZE);
+  }
+
+  @Test
+  public void testPartialChunkRead() throws Exception {
+    int len = CHUNK_SIZE / 2;
+    byte[] b = new byte[len];
+
+    chunkStream.read(b, 0 , len);
+
+    matchWithInputData(b, 0, len);
+
+    // To read chunk data from index 0 to 49 (len = 50), we need to read
+    // chunk from offset 0 to 60 as the checksum boundary is at every 20
+    // bytes. Verify that 60 bytes of chunk data are read and stored in the
+    // buffers.
+    matchWithInputData(chunkStream.readByteBuffers.get(0).toByteArray(),
+        0, 60);
+
+  }
+
+  @Test
+  public void testSeek() throws Exception {
+    seekAndVerify(0);
+
+    try {
+      seekAndVerify(CHUNK_SIZE);
+      Assert.fail("Seeking to Chunk Length should fail.");
+    } catch (EOFException e) {
+      GenericTestUtils.assertExceptionContains("EOF encountered at pos: "
+          + CHUNK_SIZE + " for chunk: " + CHUNK_NAME, e);
+    }
+
+    // Seek before read should update the ChunkInputStream#chunkPosition
+    seekAndVerify(25);
+    Assert.assertEquals(25, chunkStream.getChunkPosition());
+
+    // Read from the seeked position.
+    // Reading from index 25 to 54 should result in the ChunkInputStream
+    // copying chunk data from index 20 to 59 into the buffers (checksum
+    // boundaries).
+    byte[] b = new byte[30];
+    chunkStream.read(b, 0, 30);
+    matchWithInputData(b, 25, 30);
+    matchWithInputData(chunkStream.readByteBuffers.get(0).toByteArray(),
+        20, 60);
+
+    // After read, the position of the chunkStream is evaluated from the
+    // buffers and the chunkPosition should be reset to -1.
+    Assert.assertEquals(-1, chunkStream.getChunkPosition());
+
+    // Seek to a position within the current buffers. Current buffers contain
+    // data from index 20 to 59. ChunkPosition should still not be used to
+    // set the position.
```

Comment:
Can we also have a test case like seek to a position, and then call read which we have, and then continue with next read and see we are reading the continued data or not?

Reply:
Added a new test testSeekAndRead()